src/linux/cgroups.cpp

```diff
@@ -696,13 +696,24 @@ Try<Nothing> remove(const string& hierarchy, const string& cgroup)
   string path = path::join(hierarchy, cgroup);
 
   // Do NOT recursively remove cgroups.
-  Try<Nothing> rmdir = os::rmdir(path, false);
-  if (rmdir.isError()) {
-    return Error(
+  // TODO The retry was added as a fix for kernel bug
+  // https://lkml.org/lkml/2020/1/15/1349
+  // where calling rmdir on a seemingly empty cgroup can fail
+  // with EBUSY while some tasks are exiting.
+  auto delay = Milliseconds(1);
+  for (auto retry = 10; ;) {
+    Try<Nothing> rmdir = os::rmdir(path, false);
+    if (!rmdir.isError()) {
+      return rmdir;
+    } else if (retry > 0) {
+      os::sleep(delay);
```

Comment:
Note that `os::sleep()` is a blocking call while the `remove()` function is supposed to be non-blocking. Calling a blocking function from a non-blocking context must be avoided if possible.

If the error occurs for the `freezer` cgroup only, [Destroyer::remove](https://github.com/apache/mesos/blob/f8a3dd334934094ec44e07fa350f958d218bc78f/src/linux/cgroups.cpp#L1536-L1555) method can be modified to asynchronously retry "remove cgroup" operation.

Reply:
There is at least another code path which calls `cgroup::remove` - when the freezer subsystem isn't available: https://github.com/apache/mesos/blob/f8a3dd334934094ec44e07fa350f958d218bc78f/src/linux/cgroups.cpp#L1596

Also it seems that `cgroup::mount` has a similar hack: https://github.com/apache/mesos/blob/f8a3dd334934094ec44e07fa350f958d218bc78f/src/linux/cgroups.cpp#L575

It returns a future but can block retrying?

So unless you feel strongly about it maybe best to keep it there, since failing to remove a cgroup is pretty much a fatal error?

Also do you have a suggestion for this - I agree we should check for EBUSY, I'm not sure how to do it :).

> > Should we retry only if ::rmdir() returns EBUSY errno error?
> 
> Definitely - I wanted to do that but I'm not sure what's the best way to do it: is there a way to access errno from Try<Nothing> rmdir or can I just assume that the global errno is preserved and access it directly?

Reply:
An internal helper function can be introduced to implement non-blocking retry logic:
```
Future<Nothing> remove(const std::string& hierarchy, const std::vector<std::string>& groups)
````
[XfsDiskIsolatorProcess::initialize](https://github.com/apache/mesos/blob/f8a3dd334934094ec44e07fa350f958d218bc78f/src/slave/containerizer/mesos/isolators/xfs/disk.cpp#L973-L982) and [IOSwitchboard::_connect](https://github.com/apache/mesos/blob/f8a3dd334934094ec44e07fa350f958d218bc78f/src/slave/containerizer/mesos/io/switchboard.cpp#L695-L713) are examples of how `process::loop()` can be used to implement non-blocking retries.

The new `remove` function can be used in both places you have mentioned above where the blocking `cgroups::remove` function is called.

> we should check for EBUSY

Just call `::rmdir()` (no `os::` namespace)

Reply:
> XfsDiskIsolatorProcess::initialize and IOSwitchboard::_connect are examples of how process::loop() can be used to implement non-blocking retries.

> The new remove function can be used in both places you have mentioned above where the blocking cgroups::remove function is called.

Sorry, I don't really see how to do this in a helper function: the problem is that we need to persist state from one call to the next - e.g. the delay and current retry count - and we don't have a scope where those variables would be live.

For example if the function looked like:

```
Future<Nothing> remove(const std::string& hierarchy, const std::string& cgroup)
{
  const auto path = path::join(hierarchy, cgroup);

  auto delay = Milliseconds(1);
  auto retry = 10;
  auto result = Future<Nothing>();

  return loop(
      [&]() {
        return process::after(delay);
      },
      [&](const Nothing&) -> ControlFlow<Nothing> {
        if (::rmdir(path.c_str()) == 0) {
          return process::Break();
        } else if ((errno == EBUSY) && (retry > 0)) {
          delay *= 2;
          --retry;
          return process::Continue();
        } else {
          // If the `cgroup` still exists in the hierarchy, treat this as
          // an error; otherwise, treat this as a success since the `cgroup`
          // has actually been cleaned up.
          if (os::exists(path::join(hierarchy, cgroup))) {
            result = Failure(
              "Failed to remove directory '" + path + "': " +
              os::strerror(errno));
          }
          return process::Break();
        }
      }
    );

  return result;
}
```

The `delay`, `retry` and `result` would be reference to local variables, so invalid.
`XfsDiskIsolatorProcess::initialize` and `IOSwitchboard::_connect` are both member functions of the processes on bahalf of which they're run, so don't have this problem.

TBH I'm not really familiar with libprocess, maybe it would be much easier for you to write the patch?

Reply:
@abudnik do you have a suggestion to help this move forward?

I've applied my patch to our Mesos build we're using at work, because without it we keep having containers which don't get destroyed, and it leads to a lot of pain because we use gpu isolation and the agent then basically leaks the gpu...

As I said earlier I really don't see how to implement the non-blocking retry in a function because we have nowhere to store the variables such as `retry`, `delay`, etc.
Can you detail how you would implement it?

Reply:
Added a few minor changes to your code:
```
Future<Nothing> remove(const string& hierarchy, const string& cgroup)
{
  // NOTE: please minimize using of `auto` keyword to keep the code consistent.
  const string path = path::join(hierarchy, cgroup);

  Duration delay = Milliseconds(1);
  int retry = 10;

  // TODO: add a comment describing the problem with the kernel bug and the way
  // we retry `rmdir` operation.
  return loop(
      // NOTE: copy `delay` by value in the first lambda function.
      [=]() mutable {
        auto timeout = process::after(delay);
        delay *= 2;
        return timeout;
      },
      // NOTE: copy `retry` by value in the second lambda function.
      [=](const Nothing&) mutable -> Future<ControlFlow<Nothing>> {
        if (::rmdir(path.c_str()) == 0) {
          return process::Break();
        } else if ((errno == EBUSY) && (retry > 0)) {
          --retry;
          return process::Continue();
        } else {
          // If the `cgroup` still exists in the hierarchy, treat this as
          // an error; otherwise, treat this as a success since the `cgroup`
          // has actually been cleaned up.
          if (os::exists(path::join(hierarchy, cgroup))) {
            return Failure(
                "Failed to remove directory '" + path +
                "': " + os::strerror(errno));
          }
          return process::Break();
        }
      }
    );
}
```

Reply:
Hm, I'm still confused: in your version, you capture `retry` and `delay` in the lambdas by value, so their modifications are lost:

```
  return loop(
      // NOTE: copy `delay` by value in the first lambda function.
      [=]() mutable {
        auto timeout = process::after(delay);
        delay *= 2;
        return timeout;
      },
      // NOTE: copy `retry` by value in the second lambda function.
      [=](const Nothing&) mutable -> Future<ControlFlow<Nothing>> {
[...]
        } else if ((errno == EBUSY) && (retry > 0)) {
          --retry;
          return process::Continue();
```

I think that with this, `delay` and `retry` will *not* be updated from one iteration to the next: so basically we will keep retrying the `rmdir` every ms forever.

Reply:
`loop()` calls the same lambda object, so the state of this lambda is preserved between iterations. Here is an example of this approach (`maxBackoff` is captured by lambda): https://github.com/apache/mesos/blob/master/src/csi/v1_volume_manager.cpp#L512-L545

Reply:
Ah, I hadn't realised that `loop` actually called the lambda multiple times, I thought it was copying it every time - makes sense, and pretty neat.

I'll finish, thanks!

Reply:
Done.
I had to use another `loop` to chain the calls to `remove` - since we want to delete directories in order, because we proceed in bottom-up fashion.

I tested in the following way:
- make check
- wrote a reproducer which repeatedly starts tasks with a large memory limit and exceed it, so that they are called by the OOM killer
- check that it was failing without the fix - I've been running the same test for several hours with the fix, no failure so far
- I also checked the rmdir handling of EBUSY with strace, checking the exponential backoff
- I also tried to reduce the maximum retries to 2 to force it to fail, and checked that failures are correctly reported